Étape de sélection du nombre de dossiers ANTS

adipasquale opened the pull request:
# Contexte

[Fiche Notion : Dispositif de Recueil - Prise de RDV via l’URL pour plusieurs usagers ](https://www.notion.so/rdvs/Dispositif-de-Recueil-Prise-de-RDV-via-l-URL-pour-plusieurs-usagers-1892b05ced4180b7974ec741bbc3625a#1892b05ced4181f4988ece5c7bb5e575)

[Maquettes Figma](https://www.figma.com/design/JBirKx0mGeWZk8D8DQTw6N/Interface-Usager?node-id=556-18132&p=f&t=F8sKx6bxcipElj8B-0)

# Solution

> [!NOTE]
> Lire la PR commit par commit 

Le gros changement est l’ajout d’un get param `ants_pre_demande_numbers_count`. Il faut le passer partout pendant le parcours de prise de rdv usagers pour les motifs concernés.

# Captures d’écran

nouvelle étape|recap|pas de resultats
-|-|-
<img width="335" alt="image" src="https://github.com/user-attachments/assets/d3243d28-0245-432d-9ede-f96ef53f59f3" /> | <img width="373" alt="image" src="https://github.com/user-attachments/assets/3b7360e6-6884-45e5-98dc-a616f8506396" /> | <img width="334" alt="image" src="https://github.com/user-attachments/assets/c642044d-8d23-4c37-b5c5-2bde61dae3ac" />

état d’erreur

![image](https://github.com/user-attachments/assets/0555ae90-27ee-44b6-aec0-301cae2c8b0a)


# Changement supplémentaire

> [!NOTE]
> Ce changement correspond au dernier commit, on pourrait le déplacer dans une 2e PR

J’ai légèrement modifié le parcours de prise de RDV en passant par le moteur de recherche de créneau de l’ANTS : https://rendezvouspasseport.ants.gouv.fr/ .

Aujourd’hui, lorsque l’ANTS appelle notre API pour lister des créneaux, nous renvoyons des URLs de prise de RDV avec un paramètre `duration` configuré en fonction de la durée du motif du créneau et du nombre de documents ANTS.

Avec l’introduction du nouveau paramètre dans cette PR nous pouvons simplement renvoyer des URLs contenant un `ants_pre_demande_numbers_count`. 

Cela permet qu’une fois arrivé sur RDVSP on garde le contexte du nombre de documents et il peut être modifié sur RDVSP : <img width="416" alt="image" src="https://github.com/user-attachments/assets/3340ba36-888f-4971-bc7e-b42aa5f2f89a" />

On va aussi (je crois) pouvoir supprimer le support du param `duration` dans le parcours usager qui permet de prendre des RDV sans respecter la durée des motifs. 

J’envisage de d’abord rajouter un warning sentry pour être prévenu s’il est utilisé quelque part et sinon de le supprimer.

# Tests sur la review app

> [!WARNING]
> Les mails sont activés sur cette review app, faites attention à ne mettre que des adresses mails que vous possédez. Les sms ne sont pas activés.

Lien de résa en ligne usagers pour la mairie de sannois sur la review app : https://demo-rdv-solidarites-pr5083.osc-secnum-fr1.scalingo.io/prendre_rdv?departement=95&public_link_organisation_id=32

usager existant [patricia_duroy@demo.rdv-solidarites.fr](mailto:patricia_duroy@demo.rdv-solidarites.fr) Rdvservicepublictest1! 

vous pouvez utiliser les numéros de prédemande ANTS suivants `RDVSPUB001`, `RDVSPUB002`, `RDVSPUB003`.

Teodora-Stanki commented:
Merci Adrien, voici mes retours (principalement UI, a voir d'ailleurs si c'est le trio "parcours usager" qui doit s'en occuper plutôt) :
 
- Concernant le texte, je serai d'avis de proposer celui-ci :
"Pour prendre rendez-vous avec une mairie, **chaque personne** souhaitant une nouvelle pièce d'identité doit avoir rempli une pré-demande en ligne sur le site de l’ANTS. Faire votre pré-demande "

- Est-il possible de mettre un margin-top au CTA de 16px / 1 rem ? En effet sur mobile avec mon gros pouce je peux facilement toucher la zone du moins au lieu du cta.
<img width="1430" alt="Capture d’écran 2025-04-16 à 11 39 14" src="https://github.com/user-attachments/assets/969c31de-57a8-4dd1-b5e8-b8e1595a9a44" />

- J'ai une alerte "veuillez respecter le format requis" lorsque je selectionne 10 pré-demandes, je ne comprends pas a quoi cela correspond.
![signal-2025-04-16-120042_002](https://github.com/user-attachments/assets/b73ddc0d-81ee-4952-8fc3-5a47813586c2)

- Je n'ai pas le meme résultat que toi lorsqu'il n'y a pas de créneaux disponibles, je sais pas si c'est normal : 
![signal-2025-04-16-120152_002](https://github.com/user-attachments/assets/91c8a4ce-357b-41f6-9873-29d24a804a5b)

(... demandes sur les boutons déplacés vers #5254)

En tous cas bonne première amélioration :) 
N'hésite pas a me dire si les retours UI seraient a effectuer avec l'autre trio plutot que pour cette PR

adipasquale commented:
Merci beaucoup @Teodora-Stanki pour tes retours ! j’en ai déplacé une partie vers #5254

> Concernant le texte, je serai d'avis de proposer celui-ci :
> "Pour prendre rendez-vous avec une mairie, **chaque personne** souhaitant une nouvelle pièce d'identité doit avoir rempli une pré-demande en ligne sur le site de l’ANTS. Faire votre pré-demande "

Merci c’est fait ! Je me suis permis de: 
- retirer le début `Pour prendre rendez-vous avec une mairie` 
- et de modifier légèrement  `souhaitant une nouvelle pièce d'identité` en `souhaitant une nouvelle carte d'identité ou passeport`


> Est-il possible de mettre un margin-top au CTA de 16px / 1 rem ? En effet sur mobile avec mon gros pouce je peux facilement toucher la zone du moins au lieu du cta.

c’est fait ✅ 

au passage le champ texte contenant le nombre de pré-demandes était trop étroit en vue mobile j’ai corrigé çà.

> J'ai une alerte "veuillez respecter le format requis" lorsque je selectionne 10 pré-demandes, je ne comprends pas a quoi cela correspond.

J’avais mis en place une limite côté navigateur qui vérifiait le contenu du champ pour qu’il soit entre 1 et 9 strictement mais c’est un peu bancal cette fonctionnalité des navigateurs pour valider les champs numériques… je l’ai désactivé, tu peux maintenant écrire n’importe quoi dans ce champ et on fera la validation côté serveur

> Je n'ai pas le meme résultat que toi lorsqu'il n'y a pas de créneaux disponibles, je sais pas si c'est normal :

C’est ma faute je n’avais pas déployé la bonne version avant de vous écrire en vous demandant de tester 🤦 c’est fait. 
Tu peux accéder directement à la page d’erreur en suivant ce lien : https://demo-rdv-solidarites-pr5083.osc-secnum-fr1.scalingo.io/prendre_rdv?departement=95&service_id=6&motif_name_with_location_type=passeport-public_office&public_link_organisation_id=32&ants_pre_demande_numbers_count=9&commit=Valider

je veux bien ton avis dessus !

Teodora-Stanki commented:
Merci pour ces modifications.
J'ai pu tester la page de résultat vide. Je propose d'utiliser un autre composant que voici : 
![Capture d’écran 2025-04-16 à 16 07 12](https://github.com/user-attachments/assets/413eea0c-247c-4ad4-a6f6-6ee039ccffe6)
https://storybook.systeme-de-design.gouv.fr/?path=/docs/callout--docs&globals=viewport:lg

Pour le texte, je propose ca, mais on peut modifier si la formulation n'est pas assez concise :

**Élargissez votre recherche**
Vous pouvez prendre rendez-vous pour un passeport ou une carte d’identité dans n’importe quelle mairie en France, quel que soit votre lieu de résidence ou de travail, grâce au moteur de recherche de créneaux de l'ANTS : 
Chercher d'autres créneaux (avec le picto qui va bien comme t'as fait)


Pour le module de choix du nombre de pré demande, en regardant chez la concurrence et sur le site de l'ANTS, on propose max 6 dossiers, ce qui me parait raisonnable. On peut partir sur cette limite là je pense.

adipasquale committed:
Étape de sélection nombre dossiers ANTS

C’est le coeur de cette PR. On rajoute une étape
dans le parcours de prise de RDV pré-connexion.
On passe le nouveau paramètre partout où c’est
nécessaire.

adipasquale committed:
Ajout de specs E2E pour la nouvelle étape

adipasquale committed:
MàJ des liens de callback envoyés à l’ANTS

Il est maintenant plus utile que les personnes
arrivant du moteur de recherche de l’ANTS arrivent
avec un paramètre contenant le nombre de dossiers
à déposer plutôt que la durée du créneau souhaité.

adipasquale committed:
pur refactor E2E spec RDV Mairie

Cette PR ne change rien mais prépare pour que les
changement à venir soient plus lisibles.
J’éclate principalement des lignes longues en
plusieurs lignes

victormours commented on app/controllers/search_controller.rb:
```diff
@@ -72,7 +72,8 @@ def public_link_to_creneaux
       motif_id: motif.id,
       public_link_organisation_id: params[:public_link_organisation_id],
-      duration: params[:duration]
+      duration: params[:duration], # TODO: supprimer ce param
```
je suis assez confiant que l'intégration à l'ants était le seul endroit où on utilisait ce param 👍

adipasquale replied:
super merci, je le supprimerai dans une PR successive

victormours commented on app/javascript/components/counter-field.js:
```diff
@@ -0,0 +1,38 @@
+      event.preventDefault()
+      if (parseInt(inputElt.value) < maxValue)
+        inputElt.value = parseInt(inputElt.value) + 1
+      updateDisabled()
```
un petit détail, vu qu'on a une `{` ligne 23 mais pas ligne 33, on fera toujours le `updateDisabled()` lors de l'increment mais pas toujours pour le decrement. Je crois que c'est plutôt le `increment` qui est le bon comportement (et je suis un peu déçu que le linter js n'ai pas réagit à l'indentation fausse ligne 25 🤔 )

adipasquale replied:
merci ! c’est corrigé, en effet le bon comportement est increment

victormours commented on app/javascript/components/counter-field.js:
```diff
@@ -0,0 +1,38 @@
+      event.preventDefault()
+      if (parseInt(inputElt.value) > minValue) {
+        inputElt.value = parseInt(inputElt.value) - 1
+      updateDisabled()
```
si on se met à avoir plusieurs fois ce composant dans la page avec des `maxValue` et `minValue` différentes, on va avoir des surprises avec cette fonction qui met tous les compteurs à jour. On pourrait soit documenter qu'il vaut mieux ne pas avoir plusieurs fois ce composant, soit passer `etl` et `inputElt` à `updateDisabled`

adipasquale replied:
merci beaucoup pour tes retours sur le JS 🙇 J’ai corrigé en rajoutant un `data-controller="counter-field"` et en faisant en sorte que les appels soient scopés à ce controleur - comme stimulus.

victormours commented on app/services/concerns/users/creneaux_wizard_concern.rb:
```diff
@@ -2,13 +2,16 @@ module Users::CreneauxWizardConcern
     elsif !motif_selected?
       :motif_selection
+    elsif requires_ants_pre_demande_numbers_count_selection?
+      :ants_pre_demande_numbers_count_selection
```
👍

victormours commented on app/views/search/_ants_pre_demande_numbers_count_selection.html.slim:
```diff
@@ -0,0 +1,29 @@
+.fr-card
+  .fr-card__body
+    .fr-card__content
+      h3 Nombre de pré-demandes ANTS
```
Dans la continuité de ce que propose Téo, on pourrait même mettre `Nombre de personnes concernées` ou `Nombre de dossiers` pour être plus direct. Je pense que les gens ont plutôt tendance à se dire "on va refaire des papiers pour trois personnes" plutôt que "on va refaire des papiers pour trois pré-demandes ANTS"

adipasquale replied:
Je ne suis pas sûr à 100%. 

Je pense que c’est important à cette étape de dire clairement aux gens : arrêtez-vous, il faut faire un truc compliqué (faire tout le processus de pré-demande ANTS) avant de prendre RDV. Donc ça me paraît bien d’utiliser le vocabulaire adapté ici. On a pas mal hésité à simplifier ça en « nombre de dossiers » mais j’ai l’impression que c’est plus judicieux de réutiliser le vocabulaire de l’ANTS.

Si on met « nombre de personnes concernées » ce n’est pas hyper clair qu’il faut autant de pré-demandes que de personnes

Il y a aussi un petit cas minoritaire possible : une personne peut avoir déposé deux pré-demandes une pour un passeport et une pour une carte d’identité (au lieu d’avoir fait une seule pré-demande pour les deux). 

Je n’ai pas un avis super fort là dessus si vous pensez que c’est mieux de dire nombre de personnes pour simplifier on y va

victormours replied:
Je suis un peu influencé par la manière dont c'est présenté sur le moteur de recherche de France Titres - ANTS, où ils utilisent "nombre de personnes", mais c'est vrai qu'ils sont peut-être dans un contexte où ils ont communiqué plus clairement à l'usager qu'il faut commencer par la pré-demande.
<img width="1195" alt="Screenshot 2025-05-15 at 12 02 13" src="https://github.com/user-attachments/assets/fc4627bc-2b4e-4d43-b1c2-6ba21e2c3cf2" />

C'est un peu difficile à juger sans avoir plus d'infos du terrain, donc ça me va de merger comme ça 👍

victormours commented on app/views/search/_ants_pre_demande_numbers_count_selection.html.slim:
```diff
@@ -0,0 +1,29 @@
+    .fr-card__content
+      h3 Nombre de pré-demandes ANTS
+      p
+        |> Chaque personne souhaitant une nouvelle carte d'identité ou passeport doit avoir rempli une pré-demande en ligne sur le site de l’ANTS.
```
J'ai l'impression que la marque de l'ANTS a évolué aussi, et qu'ils utilisent plutôt le nom "France Titre", ou alors la version longue "France Titres - Agence nationale des titres sécurisés" (voir ants.gouv.fr). On pourrait rendre service à nos usagers en mettant ce nom complet plutôt que de supposer qu'il connaissent cet acronyme

adipasquale replied:
bien vu ! je fais ce changement

victormours commented on app/views/search/_ants_pre_demande_numbers_count_selection.html.slim:
```diff
@@ -0,0 +1,29 @@
+      h3 Nombre de pré-demandes ANTS
+      p
+        |> Chaque personne souhaitant une nouvelle carte d'identité ou passeport doit avoir rempli une pré-demande en ligne sur le site de l’ANTS.
+        = link_to "Faites votre pré-demande ici", "https://passeport.ants.gouv.fr/demarches-en-ligne", target: "_blank"
```
👍 trop bien d'avoir ajouté ça

victormours commented on app/javascript/components/counter-field.js:
```diff
@@ -0,0 +1,38 @@
+      updateDisabled()
+    })
+  })
+}
```
éventuellement il faudrait mettre un `updateDisabled` sur un eventlistener sur l'input aussi

adipasquale replied:
bien vu ! c’est rajouté

au passage j’ai rajouté un `inputmode=number` sur le champ pour les mobiles.

à noter qu’il n’y a pas de validation client side sur ce champ tu peux rentrer n’importe quelle valeur, lettres, chiffre hors de la limite ou pas.

victormours commented on app/views/search/_nothing_to_show.html.slim:
```diff
@@ -12,4 +12,11 @@
+          h3.fr-callout__title Élargissez votre recherche
+          p.fr-callout__text
+           |>  Vous pouvez prendre rendez-vous pour un passeport ou une carte d’identité dans n’importe quelle mairie en France, quel que soit votre lieu de résidence ou de travail, grâce au moteur de recherche de créneaux de l'ANTS :
+          = link_to "Chercher d’autres créneaux", "https://rendezvouspasseport.ants.gouv.fr/", target: "_blank", class: "fr-btn"
```
👍 trop bien d'avoir géré ce cas aussi joliment

victormours reviewed (approved):
👍 Trop bien ! J'ai quelques nitpicks sur des détails de JS et de wording pour s'harmoniser avec France Titres / ANTS, mais l'implém et les tests sont nickels. Merci aussi d'avoir été aussi attentif au détails pour réorienter les usagers vers les bonnes étapes du parcours France Titres.

adipasquale committed:
improve JS after victor review

adipasquale committed:
changer le nom de l’ANTS

adipasquale committed:
Gestion côté serveur des numéros de pré-demandes ANTS invalides

J’ai exploré plusieurs pistes sur la manière d’implémenter ces
vérifications : form object, validateur AR. J’ai aussi testé de
brancher la validation au niveau des controllers mais ce n’était
pas très satisfaisant.

adipasquale reviewed (commented):
merci beaucoup pour tes retours @victormours j’ai corrigé quasiment tout ce que tu me signalais à part le renommage de « pré-demande » en « personnes », à rediscuter si tu veux ! 

en revanche je me suis rendu compte que je ne faisais aucune validation côté serveur du nouveau param `ants_pre_demande_numbers_count`. On ne fait en effet quasiment aucune validation de format ou de cohérence des params lors de la recherche. On est très souples et ce qu’on surveille c’est : est-ce qu’il existe un créneau pour les params passés. 

Pour ce nouveau param je me suis dit que c’était quand même important de vérifier que les usagers ne passent pas des valeurs absurdes (-10 ou 50 par ex) ou alors carrément invalides (`abc` par ex). C’était nettement plus compliqué à faire que ce que je pensais 😅 voici où j’en suis arrivé et je suis curieux d’avoir ton retour 5a293aa3ce9179b7d8a2193aed9d9bd1f784a8e8 . J’ai essayé de mettre beaucoup de commentaires pour expliquer comment j’en suis arrivé là.

victormours reviewed (approved):
👏 Merci pour les mises à jour ! C'est top

victormours commented:
> Pour ce nouveau param je me suis dit que c’était quand même important de vérifier que les usagers ne passent pas des valeurs absurdes (-10 ou 50 par ex) ou alors carrément invalides (abc par ex). C’était nettement plus compliqué à faire que ce que je pensais 😅 voici où j’en suis arrivé et je suis curieux d’avoir ton retour https://github.com/betagouv/rdv-service-public/commit/5a293aa3ce9179b7d8a2193aed9d9bd1f784a8e8 . J’ai essayé de mettre beaucoup de commentaires pour expliquer comment j’en suis arrivé là.

Oui, ça me semble très raisonnable comme validation

victormours committed:
Mise à jour du mail de confirmation d'ouverture d'espace, et changement du lien de contact sur la page configuration (#5329)

* Update confirmation email

* update team meeting link for configuration page

* fix spec

* fix typo

* fix border width and trigger new build

adipasquale commented:
J’ai renommé : 
- `AntsPreDemandeNumbersCountValidator` -> `AntsPreDemandesCountValidator`
- `ants_pre_demande_numbers_count` -> `ants_pre_demandes_count`

en utilisant https://github.com/newren/git-filter-repo pour preserver les differents commits

adipasquale committed:
Merge branch 'production' into feature/ants-selection-users-count

adipasquale committed:
allow workflow dispatch in CI
